crates/telio-starcast/src/nat.rs

```diff
@@ -0,0 +1,648 @@
+            IpAddr::V4(_) => {
+                ip_packet.set_source(IpAddr::V4(self.vpeer_ipv4_addr));
+                ip_packet.set_checksum(ipv4::checksum(&ip_packet.to_immutable_ipv4()));
+            }
+            IpAddr::V6(_) => {
+                ip_packet.set_source(IpAddr::V6(self.vpeer_ipv6_addr));
+            }
+        };
+
+        let mut udp_packet =
+            MutableUdpPacket::new(ip_packet.payload_mut()).ok_or(Error::PacketTooShort)?;
+        let src_port = udp_packet.get_source();
+        let src_addr = SocketAddr::new(src_ip, src_port);
+        let mut nat_port = (self.mapper)(src_addr);
+
+        while let Some(&current_mapping) = self.mappings.peek(&nat_port) {
+            if current_mapping != src_addr {
+                nat_port = nat_port.wrapping_add(1);
+            } else {
+                break;
+            }
+        }
+
+        self.mappings.insert(nat_port, src_addr);
+
+        udp_packet.set_source(nat_port);
+
+        match dst_ip {
+            IpAddr::V4(dst_ip) => udp_packet.set_checksum(udp::ipv4_checksum(
+                &udp_packet.to_immutable(),
+                &self.vpeer_ipv4_addr,
+                &dst_ip,
+            )),
+            IpAddr::V6(dst_ip) => udp_packet.set_checksum(udp::ipv6_checksum(
+                &udp_packet.to_immutable(),
+                &self.vpeer_ipv6_addr,
+                &dst_ip,
+            )),
+        };
+
+        Ok(())
+    }
+
+    fn translate_outgoing_internal<'a, P: MutableIpPacket<'a>>(
+        &mut self,
+        packet: &'a mut [u8],
+    ) -> Result<(), Error> {
+        let mut ip_packet = P::new(packet).ok_or(Error::PacketTooShort)?;
+        if ip_packet.get_next_level_protocol() != IpNextHeaderProtocols::Udp {
+            return Err(Error::UnexpectedTransportProtocol);
+        }
+
+        let src_ip = ip_packet.get_source();
+
+        let mut udp_packet =
+            MutableUdpPacket::new(ip_packet.payload_mut()).ok_or(Error::PacketTooShort)?;
+        let dst_port = udp_packet.get_destination();
+
+        let dst_addr = self.mappings.get(&dst_port).ok_or(Error::MappingNotFound)?;
+
+        udp_packet.set_destination(dst_addr.port());
+        match (src_ip, dst_addr.ip()) {
+            (IpAddr::V4(src_ip), IpAddr::V4(dst_ip)) => {
+                udp_packet.set_checksum(udp::ipv4_checksum(
+                    &udp_packet.to_immutable(),
+                    &src_ip,
+                    &dst_ip,
+                ));
+                ip_packet.set_destination(IpAddr::V4(dst_ip));
+                ip_packet.set_checksum(ipv4::checksum(&ip_packet.to_immutable_ipv4()));
+            }
+            (IpAddr::V6(src_ip), IpAddr::V6(dst_ip)) => {
+                udp_packet.set_checksum(udp::ipv6_checksum(
+                    &udp_packet.to_immutable(),
+                    &src_ip,
+                    &dst_ip,
+                ));
+                ip_packet.set_destination(IpAddr::V6(dst_ip));
+            }
```

Comment:
You cannot get rid of those completely, because the dst_ip comes from cache. It's not known at compile time. I don't know if you've missed this part or you meant something else. Please let me know

Reply:
There could be parameterized proxy for this one as well, I guess? ;)
I'm not insisting though.

Reply:
Fixed according to your suggestions. Thanks again. Looks neat